File: 2000 98 & ME/QuickStart/aspplus/samples/webforms/ctrlauth/Templates/Repeater1.cs
using System;
using System.Collections;
using System.Web;
using System.Web.UI;

namespace TemplateControlSamples {

    public class Repeater1 : Control, INamingContainer {

        private ITemplate   _itemTemplate   = null;      
        private ICollection _dataSource     = null;  

        public ICollection DataSource {
            get {
               return _dataSource;
            }
            set {
               _dataSource = value;
            }
        }

        [
            Template(typeof(RepeaterItem))
        ]
        public ITemplate ItemTemplate {
            get {
               return _itemTemplate;
            }
            set {
               _itemTemplate = value;
            }
        }

        // override to prevent literal controls from being added as children
        public override void AddParsedSubObject(object o) {
        }

        // override to create repeated items
        protected override void CreateChildControls() {
            object o = State["NumItems"];
            if (o != null) {
               // clear any existing child controls
               Controls.Clear();

               int numItems = (int)o;
               for (int i=0; i < numItems; i++) {
                  // create item
                  RepeaterItem item = new RepeaterItem(i, null);
                  // initialize item from template
                  ItemTemplate.Initialize(item);
                  // add item to the child controls collection
                  Controls.Add(item);
               }
            }
        }

        // override to create repeated items from DataSource
        protected override void OnDataBinding(EventArgs e) {

            if (DataSource != null) {
                // clear any existing child controls
                Controls.Clear();
                // clear any previous viewstate for existing child controls
                ClearChildViewState();

                // iterate DataSource creating a new item for each data item
                IEnumerator dataEnum = DataSource.GetEnumerator();
                int i = 0;
                while(dataEnum.MoveNext()) {

                    // create item
                    RepeaterItem item = new RepeaterItem(i, dataEnum.Current);
                    // initialize item from template
                    ItemTemplate.Initialize(item);
                    // add item to the child controls collection
                    Controls.Add(item);

                    i++;         
                }

                // prevent child controls from being created again
                ChildControlsCreated = true;
                // store the number of items created in viewstate for postback scenarios
                State["NumItems"] = i;
            }
        }    
    }
}

File: 2000 98 & ME/QuickStart/aspplus/samples/webforms/ctrlauth/Templates/VB/Repeater1.vb
Imports System
Imports System.Collections
Imports System.Web
Imports System.Web.UI

Namespace TemplateControlSamples

    Public Class Repeater1VB : Inherits Control : Implements INamingContainer

        Private _itemTemplate As ITemplate = Nothing
        Private _dataSource As ICollection = Nothing 

        Public Property DataSource As ICollection
            Get
               Return _dataSource
            End Get
            Set
               _dataSource = Value
            End Set
        End Property
        
        public Property <Template(GetType(RepeaterItemVB))> ItemTemplate As ITemplate
            Get
               return _itemTemplate
            End Get
            Set
               _itemTemplate = value
            End Set
        End Property

        ' override to prevent literal controls from being added as children
        Public Overrides Sub AddParsedSubObject(O As Object)
        End Sub

        ' override to create repeated items
        Protected Overrides Sub CreateChildControls()
            Dim O As Object = State("NumItems")
            If Not (O = Nothing)
               ' clear any existing child controls
               Controls.Clear()

               Dim I As Integer
               Dim NumItems As Integer = CInt(O)
               For I = 0 To NumItems - 1
                  ' create item
                  Dim Item As RepeaterItemVB = New RepeaterItemVB(I, Nothing)
                  ' initialize item from template
                  ItemTemplate.Initialize(Item)
                  ' add item to the child controls collection
                  Controls.Add(Item)
               Next
            End If
        End Sub

        ' override to create repeated items from DataSource
        Protected Overrides Sub OnDataBinding(E As EventArgs)

            If Not DataSource Is Nothing
                ' clear any existing child controls
                Controls.Clear()
                ' clear any previous viewstate for existing child controls
                ClearChildViewState()

                ' iterate DataSource creating a new item for each data item
                Dim DataEnum As IEnumerator = DataSource.GetEnumerator()
                Dim I As Integer = 0
                Do While (DataEnum.MoveNext())

                    ' create item
                    Dim Item As RepeaterItemVB = New RepeaterItemVB(I, DataEnum.Current)
                    ' initialize item from template
                    ItemTemplate.Initialize(Item)
                    ' add item to the child controls collection
                    Controls.Add(Item)

                    I = I + 1
                Loop

                ' prevent child controls from being created again
                ChildControlsCreated = true
                ' store the number of items created in viewstate for postback scenarios
                State("NumItems") = I
            End If
        End Sub

    End Class

End Namespace